src/browser/provider/host.js

```diff
@@ -0,0 +1,103 @@
+import { assignInWith } from 'lodash';
+import { getViewportSize } from 'testcafe-browser-natives';
+
+
+export default class BrowserProviderHost {
+    constructor (providerObject) {
+        assignInWith(this, providerObject, (hostProp, providerProp, propName) => {
+            var overloadedMethodName = '$' + propName;
+
+            if (this[overloadedMethodName]) {
+                this[overloadedMethodName] = providerProp;
+                return hostProp;
+            }
+
+            return providerProp;
+        });
+
+        this.evaluators          = [];
+        this.initialized         = false;
+    }
+
+    async init () {
+        if (this.initialized)
+            return;
+
+        this.initialized = true;
+
+        await this.$init();
+    }
+
+    async dispose () {
+        if (!this.initialized)
+            return;
+
+        this.initialized = false;
+
+        await this.$dispose();
+    }
+
+    registerEvaluator (id, evaluator) {
+        this.evaluators[id] = evaluator;
+    }
+
+    // Helper
+    evalOnClient (id, func) {
+        return this.evaluators[id](`(${func.toString()})()`);
+    }
+
+    // API
+    async $init () {
+        return;
+    }
+
+    async $dispose () {
+        return;
+    }
+    /* eslint-disable no-unused-vars */
+    async openBrowser (id, descriptor, startPage) {
+        throw new Error('Not implemented!');
+    }
+
+    async closeBrowser (id, pageInfo) {
+        throw new Error('Not implemented!');
+    }
+
+    //optional
+    async listAvailableAliases () {
+        return [];
+    }
+
+    async isValidAlias (alias) {
+        return true;
+    }
+
+    async getDescriptorDefaultValues (alias) {
+        return null;
+    }
+
+    //optional
+    async resizeWindow (id, pageInfo, width, height) {
+        return;
+    }
+
+    //optional
+    async screenshotWindow (id, pageInfo, screenshotPath) {
+        return;
+    }
+
+    /* eslint-enable no-unused-vars */
+
+    async resizeWindowToFitDevice (id, pageInfo, deviceName, isPortrait) {
+        var size = getViewportSize(deviceName);
+
+        if (!size)
+            throw new Error();
```

Comment:
Not cool. I have a feeling that this logic just doesn't belong here. Ideally we should get rid of any non-overridable methods. Or move them to prototype as symbols to avoid mangling